src/distributed_copy.c

```diff
@@ -0,0 +1,857 @@
+	whereClauseList = list_make1(equalityExpr);
+
+	/* resolve hash function for parition column */
+	typeEntry = lookup_type_cache(partitionColumn->vartype, TYPECACHE_HASH_PROC_FINFO|TYPECACHE_CMP_PROC_FINFO);
+	hashFunction = &(typeEntry->hash_proc_finfo);
+
+	/* allocate column values and nulls arrays */
+	rel = heap_open(tableId, AccessShareLock);
+	tupleDescriptor = RelationGetDescr(rel);
+	columnCount = tupleDescriptor->natts;
+	columnValues = palloc0(columnCount * sizeof(Datum));
+	columnNulls = palloc0(columnCount * sizeof(bool));
+
+	/*
+	 * We create a new memory context called tuple context, and read and write
+	 * each row's values within this memory context. After each read and write,
+	 * we reset the memory context. That way, we immediately release memory
+	 * allocated for each row, and don't bloat memory usage with large input
+	 * files.
+	 */
+	tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+										 "COPY Row Memory Context",
+										 ALLOCSET_DEFAULT_MINSIZE,
+										 ALLOCSET_DEFAULT_INITSIZE,
+										 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * Construct hash table used for shardId->Connection mapping.
+	 * We can not use connection cache from connection.c used by GteConnection because
+	 * we need to establish multiple connections with each nodes: one connection per shard
+	 */
+	shardToConn = CreateShardToConnectionHash();
+
+	/* init state to read from COPY data source */
+	copyState = BeginCopyFrom(rel, copyStatement->filename,
+							  copyStatement->is_program,
+							  copyStatement->attlist,
+							  copyStatement->options);
+
+	if (copyState->binary)
+	{
+		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("Copy in binary mode is not currently supported")));
+	}
+	PG_TRY();
+	{
+		char partitionType = PartitionType(tableId);
+		Oid intervalTypeId = (partitionType == HASH_PARTITION_TYPE) ? INT4OID : partitionColumn->vartype;
+
+		/* Lock all shards in shared mode */
+		shardIntervalList = SortList(shardIntervalList, CompareTasksByShardId);
+
+		shardCount = shardIntervalList->length;
+		shardIntervalCache = palloc0(shardCount * sizeof(ShardInterval*));
+		hashTokenIncrement = (uint32) (HASH_TOKEN_COUNT / shardCount);
+		i = 0;
+
+		foreach(shardIntervalCell, shardIntervalList)
+		{
+			ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+			ShardId shardId = shardInterval->id;
+			LockShardData(shardId, ShareLock);
+			LockShardDistributionMetadata(shardId, ShareLock);
+			if (partitionType == HASH_PARTITION_TYPE)
+			{
+				int32 shardMinHashToken = INT32_MIN + (shardId * hashTokenIncrement);
+				int32 shardMaxHashToken = shardMinHashToken + (hashTokenIncrement - 1);
+				if (shardId == (shardCount - 1))
+				{
+					shardMaxHashToken = INT32_MAX;
+				}
+				if (DatumGetInt32(shardInterval->minValue) != shardMinHashToken ||
+					DatumGetInt32(shardInterval->maxValue) != shardMaxHashToken)
+				{
+					partitionType = RANGE_PARTITION_TYPE;
```

Comment:
I think the technique of checking the intervals in advance and then falling back to binary search is pretty neat. It might make sense to introduce a separate variable (like searchAlgorithm = BINARY_SEARCH), because you still need to calculate the hash value if the table is hash-partitioned and search for that.